docs/proposals/daemon-controller.md

```diff
@@ -0,0 +1,151 @@
+3. User should be able to create daemon pods that are launched on every node in the cluster by calling the REST API or Kubectl.
+  1. Add code in the daemon manager that creates daemon controller objects, loops through every node in the cluster, and adds the specified pod to the node (via a call to API server)
+  2. New nodes added to the cluster will run daemon pods. This involves adding code to daemon manager that listens for node additions, and adds all daemons to the new node
+  3. Add support for labels (for launching daemon pods in exactly those nodes with a certain label).
+    1. Add an extra field to the spec of the Daemon controller that allows users to specify a node selector. Modify the code that validates the daemon controller spec accordingly.
+    2. DaemonManager should check that a node’s label satisfies the daemon controller’s node selector before running a daemon on the node.
+    3. DaemonManager should check that a newly added node’s label satisfies the daemon controller’s node selector before running a daemon on the node.
+  4. When a node gets a new label, daemon controllers with selectors corresponding to the new label should run their daemon pods on the node. This involves adding code in the Daemon Manager that listens for changes in a node’s label, loops through the Daemon Controllers, and notifies the controllers if they need to take action.
+4. Add support for other methods (describe, stop, update, rolling-update, etc)
+5. Write documentation for Daemon Controller, and integrate the Daemon Controller into existing documentation articles.
+
+## Testing
+
+Testing policy: A milestone is only considered complete if it has been satisfactorily tested (as determined by code reviewers).
+
+Unit Tests:
+Each component will be unit tested, fakes will be implemented when necessary. For example, when testing the client, a fake API server will be used.
+
+Integration Tests:
+<WIP>
+
+End to End Tests:
+At least one end-to-end test will be implemented. The end-to-end test(s) should test that the daemon controller runs the daemon on every node, that the daemon is added to new nodes joining the cluster, that when a daemon is killed it restarts, and that the daemon controller supports standard features like ‘stop’, ‘describe’, etc. There are some e2e tests that could use a Daemon Controller (e.g. the network test). We will switch over to using the Daemon Controller for these tests.
+
+## Open Issues
+- Remove controller from name, Just use Daemon.
+- Consider using [experimental API prefix](https://github.com/GoogleCloudPlatform/kubernetes/issues/10009)
+- Rolling updates across nodes should be performed according to the [anti-affinity policy in scheduler](https://github.com/GoogleCloudPlatform/kubernetes/blob/master/plugin/pkg/scheduler/api/v1/types.go). We need to figure out how to share that configuration.
+- See how this can work with [Deployment design](https://github.com/GoogleCloudPlatform/kubernetes/issues/1743).
```

Comment:
Deployment is similar in some respects, but not the same. I don't know that there's much functionality that could be shared.